fix: rework ops.main type hints to allow different flavours (callable class)

dimaqq opened the pull request:
Merge readiness check
- [x] implementation
- [x] tests
- [x] docs
- [x] re-export some symbols

Release readiness check
- [ ] full super-tox
- [ ] jhack PR or decision to re-export symbols + small PR
- [ ] 2 Scenario PRs or decision to re-export symbols
- [ ] `canonical/hardware-observer-operator` PR

Slated for ops==2.17.0

https://warthogs.atlassian.net/browse/CHARMTECH-232

tonyandrewmeyer commented on test/test_main_invocation.py:
```diff
@@ -0,0 +1,111 @@
+    os.environ.pop('OPERATOR_DISPATCH', None)
+
+
+class IdleCharm(ops.CharmBase):
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
```
Is there any reason to use this rather than `CharmBase`?

dimaqq replied:
No good reason, I wanted to make this look like a real user charm. Would you rather I have changed this to CharmBase?

tonyandrewmeyer replied:
I think it would be simpler/cleaner, but if you like it better this way I don't object to it.

dimaqq replied:
done

tonyandrewmeyer commented on test/test_main_type_hint.py:
```diff
@@ -0,0 +1,78 @@
+
+def type_test_dummy(_arg: Callable[[Type[ops.CharmBase], bool], None]):
+    """
+    Helper to verify that
```
Seems like maybe more words were intended here?

tonyandrewmeyer commented on ops/main.py:
```diff
@@ -11,529 +11,18 @@
-from ops.jujucontext import _JujuContext
-from ops.log import setup_root_logging
-
-CHARM_STATE_FILE = '.unit-state.db'
```
Is this still importable as `from ops.main import CHARM_STATE_FILE`? It doesn't seem like it:

```python
>>> from ops.main import CHARM_STATE_FILE
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'CHARM_STATE_FILE' from 'ops.main' (/home/tameyer/scratch/dima-operator/ops/main.py)
```

dimaqq replied:
Good catch, jhack does this:

```
from ops.main import (
15 | CHARM_STATE_FILE,
16 | CharmMeta,
```

And Scenario does this.

Let's agree on Matrix if we re-export some symbols and/or what symbol set to reexport.

tonyandrewmeyer replied:
For what it's worth, I think it's ok to break Scenario (as long as we also fix it before we do an ops release with the breaking change, like we did with `_JujuContext`) but it is a warning that other things might be doing this too.

It seems unlikely that a charm would have any reason to use `CHARM_STATE_FILE` or `setup_root_logging`, but I can definitely see how someone could legitimately be wanting to use `CharmMeta` and they type `ops.main` and get it autocompleted from there.

My preference is to be overly cautious by being maximally backwards compatible (except for silly things like `from ops.main import logging`). If the rest of the team feels like we only need to do that if we can find an example where it would actually break a charm, I can live with that.

dimaqq replied:
I'm thinking it's easier to re-export a set of symbols with a deprecation comment and/or proxy function.

Release that, and take our time to update 2 major scenario releases, jhack, etc;

Later on we can clean up those deprecated bits. WDYT?

dimaqq replied:
Actually, even that's not entirely enough, consider this jhack code:

```py
import ops.main
--
182 |  
183 | ops.main.setup_root_logging = _patch_logger
```

dimaqq replied:
@PietroPasotti a penny for your thoughts!

benhoyt replied:
I think we can just import the 2 or 3 symbols from `_main` here, and include a comment about why. Probably not worth trying to formally deprecate or have a deprecation warning.

dimaqq replied:
I've decided to re-export the set of symbols that Scenario 6 uses specifically.

Rationale: many charms use Scenario and we shouldn't potentially break a whole bunch of charms if they pin dependencies asymmetrically (compatible ops, exact scenario); and to decouple Scenario 6 update from this PR in calendar time.

tonyandrewmeyer commented on ops/main.py:
```diff
@@ -11,529 +11,18 @@
+from typing import Optional, Type
 
 import ops.charm
-import ops.framework
-import ops.model
-import ops.storage
-from ops.charm import CharmMeta
-from ops.jujucontext import _JujuContext
-from ops.log import setup_root_logging
```
I worry about these (other than `_JujuContext`). No-one *should* be doing `from ops.main import CharmMeta` but as we saw with the _JujuContext PR, sometimes people are importing from the wrong place, and I think we should be careful to not break those (it's really our fault for not defining `__all__` or making them private in the first place).

I think it is ok to assume no-one is doing `from ops.main import ops`. So I think we can just import and export `CharmMeta` and `setup_root_logging` with a note that it's for backwards compatibility, just like we did in the `_JujuContext` PR.

dimaqq replied:
I've decided to re-export the set of symbols that Scenario 6 uses specifically.

Rationale: many charms use Scenario and we shouldn't potentially break a whole bunch of charms if they pin dependencies asymmetrically (compatible ops, exact scenario); and to decouple Scenario 6 update from this PR in calendar time.

tonyandrewmeyer commented on ops/main.py:
```diff
@@ -11,529 +11,18 @@
-CHARM_STATE_FILE = '.unit-state.db'
-
-
-logger = logging.getLogger()
```
For what it's worth, I'm fine with breaking people doing `from ops.main import logger`, and also things like `from ops.main import logging`.

dimaqq replied:
We could re-export symbols via new ops/main.py

Let's reach a decision on matrix.

benhoyt replied:
Yeah, I think it's reasonable to re-export symbols, but let's not get too carried away (no need to do `logger` or `os` or whatever). Just the ones people realistically might use -- though I realise that's a bit subjective.

dimaqq replied:
I've decided to re-export the set of symbols that Scenario 6 uses specifically.

Rationale: many charms use Scenario and we shouldn't potentially break a whole bunch of charms if they pin dependencies asymmetrically (compatible ops, exact scenario); and to decouple Scenario 6 update from this PR in calendar time.

tonyandrewmeyer commented on ops/_main.py:
```diff
@@ -0,0 +1,555 @@
+    Args:
+        charm_class: the charm class to instantiate and receive the event.
+        use_juju_for_storage: whether to use controller-side storage. If not specified
+            then Kubernetes charms that haven't previously used local storage and that
```
This is wrong in the current ops docs, but while we're here, could we fix this? The "Kubernetes charms" is no longer correct. It's only the case for podspec charms, not Kubernetes sidecar charms, which is the default K8s choice now.

dimaqq replied:
done

tonyandrewmeyer commented on ops/_main.py:
```diff
@@ -0,0 +1,555 @@
+
+"""Main entry point to the framework.
+
+Note that this module is callable, and calls the :func:`ops.main.main` function.
```
I think this is not correct any more?

dimaqq replied:
done

tonyandrewmeyer reviewed (approved):
Looks good to me - a few minor queries/suggestions.

I did not review any of the code in `_main.py` other than `main()`, assuming that it was unchanged from `main.py` - if that's not the case and there are tweaks there too, could you please note that in the PR description and I'd want to re-review.

Speaking of the PR description: I assume it's because there are three competing solutions in different PRs, but it's currently empty: could you please fix that?

dimaqq commented:
I've updated the docs and this is the best I managed to achieve so far:

![Screenshot 2024-08-29 at 10 31 41](https://github.com/user-attachments/assets/d971152e-5db9-400d-86fa-ab6ad54884fb)

tonyandrewmeyer commented on ops/_main.py:
```diff
@@ -0,0 +1,540 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Imeplement the main entry point to the framework."""
```
```suggestion
"""Implement the main entry point to the framework."""
```

dimaqq commented:
`py:func reference target not found: ops.main [ref.func]` I'm at a loss how to link to new top-level `ops.main()`

tonyandrewmeyer commented on ops/__init__.py:
```diff
@@ -321,3 +321,46 @@
+    class SomeCharm(ops.CharmBase):
+        ...
+
+    ops.main(SomeCharm)
```
Probably should have the `if __name__ == '__main__'` guard here?

dimaqq replied:
done

tonyandrewmeyer commented on docs/index.rst:
```diff
@@ -10,12 +10,27 @@ ops module
+   :noindex:
+
+
+_main module
+------------
+
+.. automodule:: ops._main
+   :noindex:
```
I think we probably shouldn't expose this in the docs. Having the legacy one there to point out the deprecation is good, though.

dimaqq replied:
done

dimaqq replied:
So I've hidden the module, but that also means that all of the module content, incl. `_Dispatcher` is gone from the docs.

Is that acceptable/intended, or do we want `_main` content documented, but not the `_main.main()` function?

tonyandrewmeyer replied:
We currently don't have any of the private methods/classes in the docs, so that should be fine. Most of them are undocumented anyway, and I don't think we want to be exposing them to people.

tonyandrewmeyer commented:
> `py:func reference target not found: ops.main [ref.func]` I'm at a loss how to link to new top-level `ops.main()`

Can you just put in a custom inline target and link to that instead? You'd lose the auto formatting, but at least it would be a link. Like:

```python
def main(...):
    """_`ops.main()` is the thing you want to call."""
```

> Charms should call `ops.main()`_ in order to handle the event.

dimaqq commented:
Addressed comments, new doc screenie:

![Screenshot 2024-08-29 at 14 03 14](https://github.com/user-attachments/assets/a778b36a-ee67-4ca5-8e21-1a78f2d5e3b8)

dimaqq commented:
![Screenshot 2024-08-29 at 14 04 51](https://github.com/user-attachments/assets/03c00af6-b81a-449b-80c6-0c04a72633de)

dimaqq commented:
![Screenshot 2024-08-29 at 15 09 09](https://github.com/user-attachments/assets/5fa957f1-d932-4aa9-bc9c-7541505ff672)

benhoyt commented on ops/__init__.py:
```diff
@@ -321,3 +321,47 @@
+
+.. code-block:: python
+
+    import ops
+
+
+    class SomeCharm(ops.CharmBase):
+        ...
+
+
+    if __name__ == "__main__":
+        ops.main(SomeCharm)
```
Nit: Can we remove the double blank lines? I realise we'd do this in real code (formatter might even require it), but it feels like a lot of vertical whitespace in the docs. Also, let's see if I can sneak in a change from `"` to `'`. :-)

```suggestion
    import ops

    class SomeCharm(ops.CharmBase):
        ...

    if __name__ == '__main__':
        ops.main(SomeCharm)
```

benhoyt commented on test/test_main_invocation.py:
```diff
@@ -0,0 +1,106 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
```
Nit: blank line after copyright header.

```suggestion

import os
```

dimaqq replied:
Thanks. I wonder why ruff doesn't have an opinion on that 😅

benhoyt commented on test/test_main_type_hint.py:
```diff
@@ -0,0 +1,79 @@
+import ops
+
+
+def type_test_dummy(_arg: Callable[[Type[ops.CharmBase], bool], None]):
```
This file, and how these are tested, is confusing to me. Can't we just rely on `test_main_invocation.py` to do the type checking, as Pyright is run over that too? What is this module actually testing, and how do these helpers check that?

I think it's not helped by the helper functions' docstrings being fairly vague, "Helper to verify" and "Helper for negative tests".

dimaqq replied:
I started with a single file where each pytest test had NxM things:
- {type, runtime} tests
- {positive, negative} tests
- {old, new} import style

That was compact but too complex.

So I've split it all up: separate tests, separate files. A single test file with all the individual tests would work just as well.

I can think how to make this a little better still, maybe assigning to a Protocol would be clearer?

Certainly the helper docstrings could be better 🙈 

To answer your question, I want to make sure that the type hints contain correct call signatures for all the ways that main() can be called now.

Even if we don't advertise specific call style or don't want charmers to use a specific call style, just because they can, I felt that type hints must be correct. It's OK to remove some type hints, but I would very much like to avoid accidentally having an present but incorrect type hint.

Call signature goes beyond a single call site: formal argument names, argument types for those with defaults, return type (if we had that).

It's also possible that I'm overthinking this.

dimaqq replied:
I've updated this file to slightly better style of testing and added docstrings.

benhoyt reviewed (commented):
Couple of nit comments, comment about re-exporting those few names in `main.py`, and question whether we need `test_main_type_hint.py`.

benhoyt commented:
Docs look good though -- thanks!

dimaqq commented:
I've decided to re-export the set of symbols that Scenario 6 uses specifically.

Rationale: many charms use Scenario and we shouldn't potentially break a whole bunch of charms if they pin dependencies asymmetrically (compatible ops, exact scenario); and to decouple Scenario 6 update from this PR in calendar time.

james-garner-canonical commented on test/test_main_type_hint.py:
```diff
@@ -0,0 +1,108 @@
+    ): ...
+
+
+class CallableWithoutArguments(Protocol):
```
Is this needed (`CallableWithoutArguments`)? Likewise, is `CallableWithCharmClassOnly` necessary? If a value conforms to `CallableWithCharmClassAndStorageFlag`, then (a) it can't be called without arguments and (b) it can be called with the first argument only (since the second argument has a default value).

Perhaps this could be reduced to:
```
class MainCall(Protocol):
    def __call__(
        self, charm_class: Type[ops.charm.CharmBase], use_juju_for_storage: Optional[bool] = None
    ): ...
```

The assignments to type check `main` can then be reduced to
```
def top_level_import() -> None:
    """Recommended style."""
    import ops

    main1: MainCall = ops.main
    main2: MainCall = ops.main.main   # deprecated style

def submodule_import() -> None:
    """Deprecated style."""
    import ops.main

    main1: MainCall = ops.main # type: ignore[assignment]
    # type checker limitation https://github.com/microsoft/pyright/issues/8830
    main2: MainCall = ops.main.main

def import_from_top_level_module() -> None:
    """Also ok."""
    from ops import main

    main1: MainCall = main
    main2: MainCall = main.main  # deprecated style


def import_from_submodule() -> None:
    """Deprecated style."""
    from ops.main import main

    main1: MainCall = main
```

Even if both `CallableWithCharmClassOnly` and `CallableWithCharmClassAndStorageFlag` are necessary, I'd favour removing `CallableWithoutArguments` since there isn't really a nice way do something like `with raises(AssignmentError)` for type checking afaik, so those lines are currently just `type: ignore`d anyway.

dimaqq replied:
CallableWithCharmClassOnly is the main way the charms are written.

The two-argument variety is rare, and may go away one day, as Juju storage will no longer be supported in Juju 4.

The point of WithoutArguments is to two-fold, and one could argue that it could be dropped from the test:
- as a negative test (if charm code looks like this, pyright will error out, helping charmers write code that actually works)
- as a sort-of self-test (as in to validate the entire approach of type testing)

I think that overall I went for "explicit is better than implicit".

I think that only testing against CallableWithCharmClassAndStorageFlag has this downside: imagine that one day, someone merges this change:

```diff
 class _Main:
     def __call__(
-        self, charm_class: Type[charm.CharmBase], use_juju_for_storage: Optional[bool] = None
+        self, charm_class: Optional[Type[charm.CharmBase]] = None, use_juju_for_storage: Optional[bool] = None
     ):
         return _main.main(charm_class=charm_class, use_juju_for_storage=use_juju_for_storage)
```

Let's assume that we want to catch that (a separate debate), then with this PR, pyright will produce errors in 2 places:
- test_main_invocation `ops.main() # type: ignore`
- test_main_type_hint `bad = ops.main # type: ignore`

On one hand, type tests help us prevent regressions.

On the other, they are somewhat redundant.

james-garner-canonical replied:
That makes sense. I'd forgotten about the ability of the type checker to warn on unneeded `type: ignore`s, in light of which I definitely think it's worth keeping the `CallableWithoutArguments` version for the 'self-test' reason.

I'm not sure I see how the 'negative' reason works though. Won't pyright complain to charm authors the same way about improperly calling main regardless of whether any of these tests exist or not?

I still don't fully understand the benefit of `CallableWithCharmClassOnly` either. Even if most charms are called this way, main is currently only implemented as `CallableWithCharmClassAndStorageFlag`. I guess having both tests means that if the storage flag is eventually removed, then the `CallableWithCharmClassOnly` tests are already implemented and in place, so it's just a matter of removing the old storage flag tests rather than adding new ones. Are there any other benefits I'm missing here?

In any case, don't let me hold you up with my questions, I suspect I just haven't fully gotten my head around the benefit of these tests.

james-garner-canonical replied:
Following conversation in daily sync, I'm happy to resolve this. For anyone reading, I think I'm simply questioning whether some of these checks are redundant, and the resolution is that some may be, but it's worth having them for safety, especially since ideally these tests can all be removed eventually anyway if legacy main goes away.

james-garner-canonical commented on test/test_main_type_hint.py:
```diff
@@ -0,0 +1,108 @@
+    sink.full = ops.main
+    sink.full = ops.main.main
+    sink.simple = ops.main.main
+    sink.simple = ops.main.main
```
`ops.main.main` twice, first should probably be `ops.main`

dimaqq replied:
Good catch, thanks!

james-garner-canonical commented:
I've given things a read through and it looks good to me, with the caveat that I'm still not familiar enough with the code base for that to signify a lot. I raised a couple of questions on [test_main_type_hint.py](https://github.com/canonical/operator/pull/1345/files/ede45079ec2999d414fdf4052e200bf31b743039#diff-33a501ad70958efad26f803a232189600b4a4f599d1c24011520cf020b89639d) just about the type checking tests.
